easyrest/views/menu_controler.py

```diff
@@ -103,12 +121,54 @@ def get_cats_controler(request):
     body = wrap({
         "Items": data_dict,
         "Categories": cats_list,
+        "restaurantName": rest.name,
         "isImage": False
     })
 
     return body
 
 
+@view_config(
+    route_name='menu_items',
+    request_method="POST",
+    renderer='json'
+)
+@restrict_access(user_types=['Client', 'Owner'])
+def create_restaurant_menu_item(request):
+    """
+    POST request controller. Create new restaurant menu item in database and return created item
+    """
+    menu_item_data = request.json_body
+
+    rest_id = request.matchdict['rest_id']
+    menu_type = int(request.matchdict['menu_id'])
+
+    menu_models = request.dbsession.query(
+        Menu).filter_by(rest_id=rest_id).all()
+
+    name = menu_item_data["name"]
+    description = menu_item_data["description"]
+    ingredients = menu_item_data["ingredients"]
+    img = menu_item_data["img"]
+    category = menu_item_data["category"]
+    menu_id = menu_models[menu_type - 1].id
```

Comment:
It's something out of my understanding. Could you explain what's going on here?

Reply:
![And?](https://media.giphy.com/media/hEc4k5pN17GZq/giphy.gif)

Reply:
![Still waiting](https://media.giphy.com/media/l0HlBO7eyXzSZkJri/giphy.gif)

Reply:
We have one to many relationship between menu_items and menus also one to many relationship between restaurants and menus:
restaurants -> menus (One to Many)
menus -> menu_items (One to Many)

We can create a lot of different menus for one restaurant:

So:
For creating menu item we need find restaurants first, then find menu for that restaurant.

On route `restaurant/{rest_id:\d+}/menu/{menu_id:\d+}` we making POST request with data about menu item:

   ```
    menu_item_data = request.json_body // get data from request
    rest_id = request.matchdict['rest_id']  // get restaurant id
    menu_type = int(request.matchdict['menu_id']) // get menu id where we need to create new item related for the restaurant
  

    menu_models = request.dbsession.query(
        Menu).filter_by(rest_id=rest_id).all() // from all menu models in DB took the all menu_models related by  the restaurant
    name = menu_item_data["name"]  // get data from request body here 
    description = menu_item_data["description"] // --//--
    ingredients = menu_item_data["ingredients"] // --//--
    img = menu_item_data["img"] // --//-- = empty string yet
    category = menu_item_data["category"] // --//--
   
    menu_id = menu_models[menu_type - 1].id // Here from the list of menu models we took one that we need to update where menu_type is id that was come from request.
 [menu_type  - 1] because on front end we have routes: 

- restaurant/2/menu/1

- restaurant/2/menu/2

- restaurant/2/menu/{n},
where { n } - is menu_type;

                                                                          

    category_model = request.dbsession.query(
        Category).filter_by(name=category).first() // get category model

    new_menu_item = MenuItem(name=name, description=description,
                             ingredients=ingredients, img=img, menu_id=menu_id, category_id=category_model.id, category=category_model) // create new item 

    request.dbsession.add(new_menu_item) /add to BD
    request.dbsession.flush() /commit transations

    new_menu_item_as_dict = new_menu_item.as_dict()
    new_menu_item_as_dict["category"] = category_model.name 

    request.response.status_code = 201
    return wrap(new_menu_item_as_dict, message="New menu item was successfully created") //return new item on front-end
```

Reply:
This code isn't complete yet. Will be refactored and improved.
If it still not clear for you. Ask about what part actually was misunderstood

Reply:
Well, this code really needs to be refactored :)

I believe the main problem here is bad naming. You use `menu_id` parameter here, but in real it's an index of menu in the collection. If you used the same Id stored in DB, it would avoid making extra DB query here to retrieve a list of all menus, instead getting only one record by it's primary key.

So good option here to send from frontend real `menu_id` instead of index, if frontend doesn't have it already, you need to send it with response where list of menus is sent  to frontend.